Cython/Utility/CythonFunction.c

```diff
@@ -787,36 +792,24 @@ static PyObject * __Pyx_CyFunction_CallMethod(PyObject *func, PyObject *self, Py
                 #endif
                 return result;
             }
-#if CYTHON_COMPILING_IN_LIMITED_API
-            py_name = __Pyx_CyFunction_get_name((__pyx_CyFunctionObject*)func, NULL);
-            if (!py_name) return NULL;
+            name = __Pyx_CyFunction_get_name_for_format(func);
+            if (unlikely(!name)) return NULL;
             PyErr_Format(PyExc_TypeError,
-                "%.200S() takes exactly one argument (%" CYTHON_FORMAT_SSIZE_T "d given)",
-                py_name, size);
-            Py_DECREF(py_name);
-#else
-            PyErr_Format(PyExc_TypeError,
-                "%.200s() takes exactly one argument (%" CYTHON_FORMAT_SSIZE_T "d given)",
-                f->m_ml->ml_name, size);
-#endif
-
+                __Pyx_FORMAT_NAME"() takes exactly one argument (%" CYTHON_FORMAT_SSIZE_T "d given)",
+                name, size);
+            __Pyx_DECREF_name_for_format(name);
             return NULL;
         }
         break;
     default:
         PyErr_SetString(PyExc_SystemError, "Bad call flags for CyFunction");
         return NULL;
     }
-#if CYTHON_COMPILING_IN_LIMITED_API
-    py_name = __Pyx_CyFunction_get_name((__pyx_CyFunctionObject*)func, NULL);
-    if (!py_name) return NULL;
-    PyErr_Format(PyExc_TypeError, "%.200S() takes no keyword arguments",
-                 py_name);
-    Py_DECREF(py_name);
-#else
-    PyErr_Format(PyExc_TypeError, "%.200s() takes no keyword arguments",
-                 f->m_ml->ml_name);
-#endif
+    name = __Pyx_CyFunction_get_name_for_format(func);
+    if (unlikely(!name)) return NULL;
+    PyErr_Format(PyExc_TypeError, __Pyx_FORMAT_NAME "() takes no keyword arguments",
+                 name);
+    __Pyx_DECREF_name_for_format(name);
```

Comment:
This exception generation code is used a lot. Seems worth moving it into a helper function. Or two, given the second case with the argument size. That would also centralise the "get name + decref" code and avoid additional global declarations for it.

(I considered [PyErr_FormatV](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_FormatV), but I don't think that fits nicely with additionally passing the function name.)

Reply:
I've turned it into a couple of helper macros. I was struggling to make the "size" argument version work with a function. Hopefully that's a bit clearer.

Reply:
I was thinking of something like this:
```c
static void __Pyx_CyFunction_raise_argument_count_error(PyObject *func, const char* message, Py_ssize_t size) {
    PyObject *py_name = __Pyx_CyFunction_get_name((__pyx_CyFunctionObject*)func, NULL);
    if (!py_name) return;
    PyErr_Format(PyExc_TypeError,
        "%.200S() %.50s (%" CYTHON_FORMAT_SSIZE_T "d given)",
        py_name, message, size);
    Py_DECREF(py_name);
}

static void __Pyx_CyFunction_raise_type_error(PyObject *func, const char* message) {
    PyObject *py_name = __Pyx_CyFunction_get_name((__pyx_CyFunctionObject*)func, NULL);
    if (!py_name) return;
    PyErr_Format(PyExc_TypeError,
        "%.200S() %.50s",
        py_name, message);
    Py_DECREF(py_name);
}
``